### PR TITLE
WIP: Initial work using platforms to run jobs remotely.

### PR DESCRIPTION
**Pull request by wxtim**
Replaces #3483

## Checklist

- [ ] Working against testing with a super-basic suite.
- [ ] Go through the test battery and fix tests
- [ ] Deal with subshell cases `platform = $(I can't believe it's not rose-host-select)`
- [ ] Work out what the next items on this list are

## Super basic suite for item 1
```cfg
[cylc]
    UTC mode = True

[scheduling]
    initial cycle point = 19831213T0600Z
    final cycle point = 19831213T0600Z
    max active cycle points = 1
    [[dependencies]]
        [[[PT1H]]]
            graph = exvcylcdev01-task => local-task

[runtime]
    [[SLEEP5]]
        script="""
        sleep 2
        echo -e "Antidisestablishmentarianism is a very long word. Can you spell it?"
        """ 
    [[local-task]]
        inherit = SLEEP5
    [[exvcylcdev01-task]]
        inherit = SLEEP5
        platform = exvcylcdev01
```

### CODE REVIEW

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=5067467&bid=16673804)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=5067467&bid=16673804)

File: cylc/flow/hostuserutil.py [hunk 1]
```diff
@@ -209,6 +209,25 @@ def is_remote_user(self, name):
...
+        to the current host.
+        
+        Return False if name is None.
+        Return True if host is unknown.
+
+        Todo:
+            Should this fail miserably if some hosts are remote and some are
+            not?
+        """
+        #breakpoint(header="inside is_remote_platform")
```
Comment (wxtim) [inline cylc/flow/hostuserutil.py]: Will need to come out, eventually.

File: cylc/flow/hostuserutil.py [hunk 2]
```diff
@@ -248,6 +267,9 @@ def is_remote(host, owner):
     """Shorthand for HostUtil.get_inst().is_remote(host, owner)."""
     return HostUtil.get_inst().is_remote(host, owner)
 
+def is_remote_platform(platform):
+    """Shorthand for HostUtil.get_inst().is_remote(host, owner)."""
```
Comment (wxtim) [inline cylc/flow/hostuserutil.py]: Make this correct

File: cylc/flow/scheduler.py [hunk 3]
```diff
@@ -1654,11 +1658,15 @@ def run(self):
             # PROCESS ALL TASKS whenever something has changed that might
             # require renegotiation of dependencies, etc.
             if self.should_process_tasks():
+                LOG.debug("HI")
```
Comment (wxtim) [inline cylc/flow/scheduler.py]: Debugging only

File: cylc/flow/cfgspec/globalcfg.py [hunk 4]
```diff
@@ -148,74 +150,74 @@
...
-                    'out viewer': [VDR.V_STRING],
-                    'err viewer': [VDR.V_STRING],
-                    'job name length maximum': [VDR.V_INTEGER],
-                    'execution time limit polling intervals': [
-                        VDR.V_INTERVAL_LIST],
-                },
-            },
-        },
-    },
+    # # task
```
Comment (wxtim) [inline cylc/flow/cfgspec/globalcfg.py]: @dpmatthews Surely we want to allow hosts in the suite config, but they can go in the global?

File: cylc/flow/cfgspec/globalcfg.py [hunk 5]
```diff
@@ -377,85 +379,163 @@ def load(self):
...
-                    break
-        modify_dirs = False
-        if host_key is not None:
-            # entry exists, any unset items under it have already
-            # defaulted to modified localhost values (see site cfgspec)
-            value = cfg['hosts'][host_key][item]
+        # self._transform()
+
+    @staticmethod
+    def get_platform_item_for_job(job_conf, item):
```
Comment (wxtim) [inline cylc/flow/cfgspec/globalcfg.py]: Not sure whether this is __really__ necessary

File: cylc/flow/cfgspec/globalcfg.py [hunk 6]
```diff
@@ -377,85 +379,163 @@ def load(self):
...
                 # Replace with ~owner for direct access via local filesys
                 # (works for standard cylc-run directory location).
                 if owner_home is None:
-                    owner_home = os.path.expanduser('~%s' % owner)
+                   owner_home = os.path.expanduser('~%s' % owner)
                 value = value.replace(self._HOME, owner_home)
-        if item == "task communication method" and value == "default":
-            # Translate "default" to client-server comms: "zmq"
-            value = 'zmq'
+                value = value.replace('$HOME', owner_home)
```
Comment (wxtim) [inline cylc/flow/cfgspec/globalcfg.py]: Pretty sure this isn't quite right yet.

File: cylc/flow/job_file.py [hunk 7]
```diff
@@ -118,9 +118,9 @@ def _check_script_value(value):
         return False
 
     @staticmethod
-    def _get_host_item(job_conf, key):
+    def _get_platform_item(job_conf, key):
         """Return host item from glbl_cfg()."""
-        return glbl_cfg().get_host_item(
+        return glbl_cfg().get_platform_item(
```
Comment (wxtim) [inline cylc/flow/job_file.py]: What on earth am I doing here?

**Comment by datamel** [comment]
Checked out the branch. Running a suite, it is getting stuck on `remote_tidy` (taskj_remote_mgr.py). 

If you look at line 252 : `for (host, owner), init_with_contact in self.remote_init_map.items():` 
Looks like remote_init_map is populated with just platform, not host,owner combo. But when called needs host,owner. 
I get the error: `File "/net/home/h01/mhall/cylc-flow/cylc/flow/task_remote_mgr.py", line 251, in remote_tidy
	    procs = {}
	ValueError: too many values to unpack (expected 2)`
My guess...need to decide which should remote_init_map hold. I'll have another play with it and maybe give you another call later @wxtim.

**Comment by wxtim** [comment]
Hi @datamel  - I'm pretty sure that the decision for `remote_init_map` should be holding platforms since we are trying to remotely initiate platforms. Interestingly the fact that you got this error means that you've already got further in running a suite than I have with my test setup. Just to be on the safe side I copied your test suite, but mine still gets stuck in the same place, which is very odd - I'm going to double check my test setup.

**Pull request closed without merging**